lib/instrumentation/http.js

```diff
@@ -20,47 +20,85 @@ let instrumentHTTP = (http, opts = {}) => {
     // node 10+ supports a three arg form.  Instead of writing both implementations,
     // dynamically switch on arg types.  I _hate_ code that does this, but it's easier
     // and cleaner than the alternative.
-    return function (_url, options, cb) {
+    //
+    // The http.request function can be called either as:
+    // http.request(options[, callback])
+    // or
+    // http.request(url[, options][, callback])
+    // So we can have 1-3 args to work with. For more details, see:
+    // http://nodejs.org/api/http.html#http_http_request_url_options_callback
+    return function (..._args) {
       if (!api.traceActive()) {
-        return original.apply(this, Array.from(arguments));
+        return original.apply(this, Array.from(_args));
       }
 
+      let _url;
       let combinedOptions = {};
       let callback;
 
-      if (typeof _url === "string") {
-        combinedOptions = Object.assign({}, url.parse(_url)); // node 10 doesn't use url.parse (except as a fallback).  figure out how to handle that?
-        delete combinedOptions.href;
-        if (typeof options === "object") {
-          Object.assign(combinedOptions, options);
-          // we need to fix up `host` after this (if it wasn't specified in options), since the port might have changed.
-          if (!options.host) {
-            combinedOptions.host = `${combinedOptions.hostname}:${combinedOptions.port}`;
+      if (_args.length >= 3) {
+        // We have _url, options, and callback if there are three args
+        // If we have more, someone's doing something *weird* but account for it anyway
+        _url = _args[0];
+        combinedOptions = _args[1];
+        callback = _args[2];
+      } else if (_args.length === 2) {
+        // With 2 _args passed in, we can have the following combinations:
+        // options, callback
+        // _url, callback
+        // _url, options
+
+        // First, determine if we have a callback or not, which can only be the second arg
+        if (typeof _args[1] === "function") {
+          callback = _args[1];
+
+          // Now that we know we have a callback, the first arg must be a url or options.
+          // Check if the first arg is a string (a url) or an object (options, which can confusingly be a URL object too)
+          if (typeof _args[0] === "string") {
+            // Case _url, callback
+            _url = _args[0];
+          } else {
+            // Case options, callback
+            // This could techincally be either a plain object or a URL object
+            combinedOptions = _args[0];
           }
-          callback = cb;
         } else {
-          callback = options;
+          // Since we don't have a callback, the first arg must be a url, the second options
+          _url = _args[0];
+          combinedOptions = _args[1];
+        }
+      } else if (_args.length === 1) {
```

Comment:
I like that we have this explicitly; it was perfectly find in the previous code, just implicit in whether or not `callback` was ever defined.